### PR TITLE
[FLINK-17159] Use testcontainers for Elasticsearch ITCases

### PR DESCRIPTION
**Pull request by aljoscha**
This should harden the tests because we now use random ports for the ES
nodes. I have a suspicion that we had clashes before, which were causing
test instability.

Additionally, this allows us to get rid of quite some of our own code
for setting up an ES environment. Once we drop ES5 we can additionally
drop `ElasticsearchResource` and `EmbeddedElasticsearchNodeEnvironment`.

## Brief change log

- replace use of `ElasticsearchResource` by testcontainers
- we cannot do that for ES 5 because there is no docker container for it

## Verifying this change

Covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

### CODE REVIEW

File: flink-connectors/flink-connector-elasticsearch5/src/test/java/org/apache/flink/streaming/connectors/elasticsearch5/ElasticsearchSinkITCase.java [hunk 1]
```diff
@@ -37,6 +40,21 @@
  */
 public class ElasticsearchSinkITCase extends ElasticsearchSinkTestBase<TransportClient, InetSocketAddress> {
 
+	protected static final String CLUSTER_NAME = "test-cluster";
+
+	@ClassRule
+	public static ElasticsearchResource elasticsearchResource = new ElasticsearchResource(CLUSTER_NAME);
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch5/src/test/java/org/apache/flink/streaming/connectors/elasticsearch5/ElasticsearchSinkITCase.java:46]: Can't we use it with ES5? (no docker image?)

File: flink-connectors/flink-connector-elasticsearch6/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch6DynamicSinkITCase.java [hunk 2]
```diff
@@ -61,7 +65,20 @@
 public class Elasticsearch6DynamicSinkITCase {
 
 	@ClassRule
-	public static ElasticsearchResource elasticsearchResource = new ElasticsearchResource("es-6-dynamic-sink-tests");
+	public static ElasticsearchContainer elasticsearchContainer =
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch6/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch6DynamicSinkITCase.java:68]: nit: the linebreaks look weird (ditto on other container fields).

**Review by AHeise — COMMENTED** [review]
❤️ . Let's wait for AZP.

File: flink-connectors/flink-connector-elasticsearch6/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch6DynamicSinkITCase.java (hunk 2 above)
Comment (aljoscha) [reply flink-connectors/flink-connector-elasticsearch6/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch6DynamicSinkITCase.java:68]: It's the Google Java Style, which is a subset (or superset depending on where you're coming from 😅) of our (very relaxed) coding style so I prefer that one.

**Comment by echauchot** [comment]
@aljoscha I can take a look as well if you want as I coded the Apache Beam connectors for ES.

**Comment by aljoscha** [comment]
@echauchot Sure, but this is just touching the tests as little. 😅

**Review by echauchot — COMMENTED** [review]
I notice that the flink community created a whole test environement for embedded Elasticsearch (_EmbeddedElasticsearchNodeEnvironment_) when official elastic _EsIntegTestCase_ existed (org.elasticsearch.test:framework). This framework adds all that is needed for embedding an ES instance and adds randomization for local encodings etc... But anyway this framework was difficult to use and I see that you plan on replacing your home made one by testContainers which seems very good to me. 

_ElasticsearchResource_ is now used only for ES5 and only to get the ES client. What I'm wondering is why for ES5 we cannot use testContainers: testcontainers does not seem to have restriction on ES versions for containers and official elastic docker image is available for ES 5.x: https://www.docker.elastic.co/r/elasticsearch/elasticsearch:5.6.9 even https://www.docker.elastic.co/r/elasticsearch/elasticsearch:5.3.3

Why not create the client with _PreBuiltTransportClient_ and testContainers like for the other ES versions? Thus you could
get rid of _ElasticsearchResource_ and _EmbeddedElasticsearchNodeEnvironment_ no ?

**Comment by aljoscha** [comment]
I was trying to do that but there is no official `elasticsearch-oss` image for 5.x versions provided by elastic: https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html.

I'm still seeing if we can somehow make it work.

**Comment by bsideup** [comment]
@aljoscha FYI 5.x images are available on Docker Hub:
https://hub.docker.com/layers/elasticsearch/library/elasticsearch/5.6.16-alpine/images/sha256-5e2dac90ba245b61ef764a44b811cb79f3acda1e8e8ee370a49c0cb5dd70f537?context=explore

**Comment by aljoscha** [comment]
@bsideup These images don't work out-of-box with testcontainers. I spent an hour trying to make it work with the non-elastic images but don't want to spend more time because I'm hoping we'll deprecated ES5 soon.

**Comment by echauchot** [comment]
> @bsideup These images don't work out-of-box with testcontainers. I spent an hour trying to make it work with the non-elastic images but don't want to spend more time because I'm hoping we'll deprecated ES5 soon.

But the images I sent above are the official Elastic ones for 5.x so they should work

**Comment by aljoscha** [comment]
They are different in what configuration they allow from the outside, so things that testcontainers expects to work don't work for some reason. For reference, these are the ES images from the ES docker repo: https://www.docker.elastic.co/r/elasticsearch. But the `5.3.3` image from there also doesn't work out-of-box with what I'm trying to do. I suggest to merge this as is once CI is green and someone else can try and make it work for ES5 if they want to. But I think it's not worth the effort since I'm hoping we can deprecate/remove ES5 soon.

**Comment by echauchot** [comment]
> They are different in what configuration they allow from the outside, so things that testcontainers expects to work don't work for some reason. For reference, these are the ES images from the ES docker repo: https://www.docker.elastic.co/r/elasticsearch. But the `5.3.3` image from there also doesn't work out-of-box with what I'm trying to do. I suggest to merge this as is once CI is green and someone else can try and make it work for ES5 if they want to. But I think it's not worth the effort since I'm hoping we can deprecate/remove ES5 soon.

Ok thanks for these precisions. I agree spending too much time on making it work for ES5 might not be useful as the end of life of the last ES5.x was 2019-03-11 and it is no more maintained by elastic since version 7.0 is out. LGTM for merging once CI is green. I saw no thread about deprecating/removing ES5 support in the ML so I'll create one.

**Comment by dawidwys** [comment]
@echauchot There were multiple such threads ;)
* https://lists.apache.org/thread.html/r0baa1ffbc12b0f3d9312d3145091b4df4298112e4eecb7687d89fee3%40%3Cuser.flink.apache.org%3E
* https://lists.apache.org/thread.html/r3144c1417d577ac165f9ae1f1466686312384965c4afeaf7c7a6b38f%40%3Cdev.flink.apache.org%3E

**Comment by echauchot** [comment]
@dawidwys thanks for the links, I did not found the threads, no need to create a new one then :)

**Review by echauchot — APPROVED** [review]
LGTM when CI is green

**Comment by aljoscha** [comment]
Merged!

**Pull request closed without merging**